tests/test_formula.py

```diff
@@ -0,0 +1,718 @@
+        pytest.param("num_1 + cat_1 * num_2 * cat_2", id="mixed"),
+        pytest.param("{np.log(num_1)} + {num_in_scope * num_2}", id="functions"),
+        pytest.param("{num_1 * num_in_scope}", id="variable_in_scope"),
+        pytest.param("bs(num_1, 3)", id="spline"),
+        pytest.param(
+            "poly(num_1, 3, raw=True) + poly(num_2, 3, raw=False)", id="polynomial"
+        ),
+        pytest.param(
+            "C(num_1)",
+            id="convert_to_categorical",
+        ),
+        pytest.param(
+            "C(cat_1, spans_intercept=False) * cat_2 * cat_3",
+            id="custom_contrasts",
+        ),
+    ],
+)
+def test_names_against_pandas(df, formula, ensure_full_rank):
+    num_in_scope = 2  # noqa
+    model_df = formulaic.model_matrix(formula, df, ensure_full_rank=ensure_full_rank)
+    model_tabmat = tm.from_formula(formula, df, ensure_full_rank=ensure_full_rank)
+    assert model_tabmat.model_spec.column_names == model_df.model_spec.column_names
+    assert model_tabmat.model_spec.column_names == tuple(model_df.columns)
+
+
+FORMULAIC_TESTS = {
+    # '<formula>': (<full_rank_names>, <names>, <full_rank_null_names>, <null_rows>)
+    "a": (["Intercept", "a"], ["Intercept", "a"], ["Intercept", "a"], 2),
+    "A": (
+        ["Intercept", "A[T.b]", "A[T.c]"],
+        ["Intercept", "A[T.a]", "A[T.b]", "A[T.c]"],
+        ["Intercept", "A[T.c]"],
+        2,
+    ),
+    "C(A)": (
+        ["Intercept", "C(A)[T.b]", "C(A)[T.c]"],
+        ["Intercept", "C(A)[T.a]", "C(A)[T.b]", "C(A)[T.c]"],
+        ["Intercept", "C(A)[T.c]"],
+        2,
+    ),
+    "A:a": (
+        ["Intercept", "A[T.a]:a", "A[T.b]:a", "A[T.c]:a"],
+        ["Intercept", "A[T.a]:a", "A[T.b]:a", "A[T.c]:a"],
+        ["Intercept", "A[T.a]:a"],
+        1,
+    ),
+    "A:B": (
+        [
+            "Intercept",
+            "B[T.b]",
+            "B[T.c]",
+            "A[T.b]:B[T.a]",
+            "A[T.c]:B[T.a]",
+            "A[T.b]:B[T.b]",
+            "A[T.c]:B[T.b]",
+            "A[T.b]:B[T.c]",
+            "A[T.c]:B[T.c]",
+        ],
+        [
+            "Intercept",
+            "A[T.a]:B[T.a]",
+            "A[T.b]:B[T.a]",
+            "A[T.c]:B[T.a]",
+            "A[T.a]:B[T.b]",
+            "A[T.b]:B[T.b]",
+            "A[T.c]:B[T.b]",
+            "A[T.a]:B[T.c]",
+            "A[T.b]:B[T.c]",
+            "A[T.c]:B[T.c]",
+        ],
+        ["Intercept"],
+        1,
+    ),
+}
+
+
+class TestFormulaicTests:
+    @pytest.fixture
+    def data(self):
+        return pd.DataFrame(
+            {"a": [1, 2, 3], "b": [1, 2, 3], "A": ["a", "b", "c"], "B": ["a", "b", "c"]}
+        )
+
+    @pytest.fixture
+    def data_with_nulls(self):
+        return pd.DataFrame(
+            {"a": [1, 2, None], "A": ["a", None, "c"], "B": ["a", "b", None]}
+        )
+
+    @pytest.fixture
+    def materializer(self, data):
+        return TabmatMaterializer(data)
+
+    @pytest.mark.parametrize("formula,tests", FORMULAIC_TESTS.items())
```

Comment:
What about moving these tests to a separate file `test_materializer.py` or similar?